Fix put FSM race condition (#1526)

ian-mi opened the pull request:
Cannot rely on a gen_fsm timeout event being sent even when the timeout
is set to zero. In particular, another message or event can preempt and
unset the timeout if the handler of that event does not reset it. Rather
than trying to save and restore the timeout state, manually send a
timeout event using gen_fsm:send_event.

ian-mi committed:
Handle empty exchange history

When compute_exchange_info is called on an index with no previous
exchanges return undefined for the Recent and LastAll timestamps.

Commit:
Backport of https://github.com/basho/riak_kv/pull/1520

ian-mi committed:
Fix put FSM race condition (#1526)

Cannot rely on a gen_fsm timeout event being sent even when the timeout
is set to zero. In particular, another message or event can preempt and
unset the timeout if the handler of that event does not reset it. Rather
than trying to save and restore the timeout state, manually send a
timeout event using gen_fsm:send_event.

fadushin committed:
Pass {RObj, no_old_object} into (yokozuna) INDEX call to comply with new API.  Accidentally overlooked this line in the back port from the develop-2.2 branch.

fadushin committed:
Fixed dialyzer error, and in the process we seem to have spotted a but with ensemble handoff, which could theoretically cause a vnode due to a inability to match a function clause for raw_put

fadushin committed:
Added .thumbs YAML

thumbot committed:
Merge pull request #1557 from basho/fd-backport-2.2

Thumbs Git Robot Merge.

fadushin committed:
Locked dependency on bitcask

bsparrow435 committed:
Bring index_hashtree fixes from 2.2 to 2.0 (#1574)

* Don't update trees before deleting them from disk.

Allow the case where we're about to destroy the hashtree to simply
close,
without going through a (sometimes very lengthy) update process first.
Only
used from `destroy_trees`. This condition was caught by a 4-day
performance
test, and is not generally reproducible without having tons of data, on
the order of 10s of millions of key/value pairs per partition.
* Handle potential errors in vnode fold request call when rebuilding.

`fold_fun`'s call to `riak_core_master:sync_command` ignored the result
of the call. It is possible for `sync_command` to return {error,
overload}
if the vnode is overloaded. Rather than ignore the result, make sure to
handle non-success results by failing the tree build.
* Make fold read new build limit after each sleep in `maybe_throttle_build`.

Before, the build limits (size and wait time) could only be changed
before a build started. Once it did, you were stuck with those settings.
If the vnode was large, you may want to decrease the wait or increase
the size, but we wouldn't pick it up until the rebuild was over.

ian-mi committed:
Merge pull request #1576 from basho/bjs-riak-3062-2.0.8

Handle empty exchange history

nickelization commented:
@ian-mi did you mean to make this a PR from develop-2.2 to 2.0? Given the title and description, I'm guessing maybe you meant to create this from a different branch where you just backported the race condition fix?

bsparrow435 committed:
Always drop dots from CRDTs, not just on merge (#1575)

Prior to this change, the first write of a new CRDT would end up with a
dot in its metadata, since dots were only being dropped on merge. This
could potentially lead to sibling CRDTs on disk, since our merge code
assumes CRDTs will never have dots.

Although we should never have sibling values for CRDTs present on disk,
this bug did not normally cause problems, because using the `fetch_type`
client functions would still cause the values to be merged prior to
returning them to the client. However, doing a basic `get` could cause
the siblings to be exposed to the client directly.

This was the issue that has been causing the `test_hll` test to
intermittently fail. The issue can be consistently reproduced using the
new `verify_no_datatype_siblings` test, and applying this fix causes
that test to pass.

Adding the new call to `drop_the_dot` is needed in order to fix this
issue, but removing the old call is not mandatory. However, if we remove
the old call then we will run into problems with existing CRDTs on disk
that already have dots present, so I'm leaving the old call in place to
deal with any such scenarios.

ian-mi commented:
No, this is a forward merge of the fix which was made against 2.0. More commits have been added to 2.0, however, so maybe this PR can be replaced by a forward merge of 2.0.

bsparrow435 committed:
Switch to 1.7 branch for bitcask

Use rebar lockfile to lock in version based on most recent commit.

bsparrow435 committed:
Move bitcask down the dependency tree to fix a cuttlefish issue

bsparrow435 commented:
Closing, will be forward ported in a seperate PR.
